### ISSUE TITLE
Investigate how to disable buttons

### ISSUE THREAD

**Issue by LiHaoTan** [TEAMMATES/teammates#8462]
Following the discussion in #8377  we are going to replace the `disabled` attribute with `.disabled`. While that is not sufficient to disable buttons, more work than anticipated is required for a proper solution.

Replacing `a` with `button` is purportedly a simpler solution so we don't have to consider things like placeholder anchors, `aria-disabled`, `role="button"`.

However, "direct link buttons" are trickier because we either need to add a click handler or keep them as anchor tags (and we are back to considering the complicated scenario we just wanted to avoid).

Thus it appears buttons and anchors have overlapping behavior that we want and we may need to swap one out for the other sometimes.

The question is why there isn't a semantic element that allows for the overlapping behaviors.

One possible reason is that due to the way the web has evolved, web applications have become more complicated and the specifications have to probably err on the safe side to ensure semantics don't change. Here's one discussion: https://discourse.wicg.io/t/adding-a-disabled-attribute-to-a-s/1116/6

Further details: https://github.com/TEAMMATES/teammates/pull/8377#issuecomment-366134168

This also implies if we want to improve the accessibility of TEAMMATES we also need to consider this issue.

**Comment by Crphang** [comment]
@LiHaoTan going through the guide on properly disabling `a`, I'm thinking whether making `button` work like `a` would be more complicated and messy because we need to add a different click handler for every button.

The discussion given above is very meta, discussing why links should even be disabled and it's differences with placeholder text but might not resolve our dilemma practically. A possibility to properly set our own standard for a overlapped behavior between `button` and `a` could be defining our own [web component](https://developers.google.com/web/fundamentals/web-components/customelements) which might require some resource and might have other compatibility issues.

**Comment by LiHaoTan** [comment]
> button work like a would be more complicated and messy because we need to add a different click handler for every button.

I don't think the intention was ever to make a button work like an anchor. The only thing the button needs from anchor is the `href` attribute.

Adding a click handler for every button is the modern way for web applications (AJAX for everything). Of course that poses its own problems such as unobtrusive Javascript and progressive enhancement, etc.

This implies also that this issue can also be resolved if we get to migrating to using a framework before we address this problem.

>  our own web component

It appears to me that just for disabling a button introducing web components introduces too much more work?

> The discussion given above is very meta, discussing why links should even be disabled

I think we don't have the resources to discuss whether links should be disabled? I'm saying that the web simply does not allow links to be disabled, whether we agree with it is another thing (the meta part you are talking about that I don't think we should discuss).

However, many people, including Bootstrap (the framework we are using) recognise the need for `a` tags behaving like buttons in certain situations. The appropriate use of anchors is still solely up to us because Bootstrap did not enforce any standards on that.

---

What I'm then saying is that there's no current easy way to disable links, and if we want to disable "buttons", we should use `button`.

If we instead choose to go with the `a` route, then we are altering the semantics of `a` and need to justify why we need to do that. For example, we really want users to be able to see the links on a "button", and want to continue treating it as a link not a button although it appears visually like a button. Of course, if we do this we make accessibility harder for ourselves.

So maybe another point to do is is to decide the direction to take. I prioritized accessibility when I made my points, and if we decide that that's not important then perhaps buttons appearing disabled is good enough.
